xformers/components/attention/favor.py

```diff
@@ -109,33 +110,28 @@ def _maybe_promote(x: torch.Tensor) -> torch.Tensor:
     def _causal_attention(
         k_prime: torch.Tensor, q_prime: torch.Tensor, v: torch.Tensor
     ) -> Tuple[torch.Tensor, torch.Tensor]:
-        # TODO(@lefaudeux): Rewrite as suggested in
-        # https://github.com/fairinternal/xformers/pull/61#discussion_r628420093
         # Algorithm 1 in the paper
-        ref_v = torch.ones_like(v[:, 0, :].unsqueeze(1))
+        ref_v = torch.ones_like(v.unsqueeze(2))  # BATCH x SEQ x 1 x EMB
```

Comment:
this is a memory hog, that's the issue with this approach @fmassa, but the older state was not actually trainable (this was not being tested.. fixed now), so I guess that it's better than before and improving on that could be a follow up

Reply:
see https://github.com/facebookresearch/xformers/issues/105 for tracking